iroh/src/magicsock.rs

```diff
@@ -1729,6 +1727,71 @@ enum DiscoBoxError {
     Parse(anyhow::Error),
 }
 
+/// Creates a sender and receiver pair for sending datagrams to the [`RelayActor`].
+///
+/// These includes the waker coordination required to support [`AsyncUdpSocket::try_send`]
+/// and [`quinn::UdpPoller::poll_writable`].
+fn relay_datagram_sender() -> (
+    RelayDatagramSendChannelSender,
+    RelayDatagramSendChannelReceiver,
+) {
+    let (sender, receiver) = mpsc::channel(256);
+    let waker = Arc::new(AtomicWaker::new());
+    let tx = RelayDatagramSendChannelSender {
+        sender,
+        waker: waker.clone(),
+    };
+    let rx = RelayDatagramSendChannelReceiver { receiver, waker };
+    (tx, rx)
+}
+
+/// Sender to send datagrams to the [`RelayActor`].
+///
+/// This includes the waker coordination required to support [`AsyncUdpSocket::try_send`]
+/// and [`quinn::UdpPoller::poll_writable`].
+#[derive(Debug, Clone)]
+struct RelayDatagramSendChannelSender {
+    sender: mpsc::Sender<RelaySendItem>,
+    waker: Arc<AtomicWaker>,
+}
```

Comment:
I guess this works, assuming there's never more than one `RelayDatagramSendChannelSender` registering for wakeup concurrently (and I think that's the case. Technically MagicSock is available via `&self`, but quinn is just lying with its type signature in `AsyncUdpSocket` there, and it's practically `&mut self`).

Any reason for doing this instead of trying to use [`PollSender`](https://docs.rs/tokio-util/latest/tokio_util/sync/struct.PollSender.html)?
I guess there's a chance that `PollSender::send_item` panics, if the `poll_reserve` doesn't work out as planned, but otherwise?

Reply:
I didn't know about `PollSender`...  But I don't think it helps much: all it's methods are all `&mut self` and the sequencing of `poll_reserve` & `send_item` is kind of tricky when spread across two methods.  Sending a permit around seems favourable over maybe panicking.

Reply:
Regarding the first issue again: You're sharing a single `AtomicWaker` for all tasks that might want to be woken up when there's space again. There *are* multiple tasks involved, because quinn will `try_send` on the same socket for multiple connections concurrently.
So, I don't think this code is correct.

Reply:
This is different from the `RelayDatagramRecvQueue`, because that's only accessed from a single task IIUC: The endpoint driver.

Reply:
As [discussed in discord](https://discord.com/channels/949724860232392765/1293936403733807165/1319348835952099420): It's not *this PR* that's introducing this bug, this PR is mostly refactoring the way it worked before, so let's not have my concern block this PR from moving on.